diktat-rules/src/main/kotlin/com/saveourtool/diktat/ruleset/rules/chapter4/NullChecksRule.kt

```diff
@@ -72,32 +77,59 @@ class NullChecksRule(configRules: List<RulesConfig>) : DiktatRule(
-                    KtTokens.EXCLEQ, KtTokens.EXCLEQEQEQ ->
-                        warnAndFixOnNullCheck(
-                            condition,
-                            isFixable(node, false),
-                            "use '.let/.also/?:/e.t.c' instead of ${condition.text}"
-                        ) {
-                            fixNullInIfCondition(node, condition, false)
-                        }
+                    KtTokens.EXCLEQ, KtTokens.EXCLEQEQEQ -> false
                     else -> return
                 }
+
+                val (_, elseCodeLines) = getThenAndElseLines(node, isEqualToNull)
+                val (numberOfStatementsInElseBlock, isAssignmentInNewElseBlock) = getInfoAboutElseBlock(node, isEqualToNull)
+
+                // if `if-else` block inside `init` or 'run', 'with', 'apply' blocks and there is more than one statement inside 'else' block, then
+                // we don't have to make any fixes, because this leads to kotlin compilation error after adding 'run' block instead of 'else' block
+                // read https://youtrack.jetbrains.com/issue/KT-64174 for more information
+                if (isShouldBeFixed(node, elseCodeLines, numberOfStatementsInElseBlock, isAssignmentInNewElseBlock)) {
+                    warnAndFixOnNullCheck(
+                        condition,
+                        isFixable(node, isEqualToNull),
+                        "use '.let/.also/?:/e.t.c' instead of ${condition.text}"
+                    ) {
+                        fixNullInIfCondition(node, condition, isEqualToNull)
+                    }
+                }
             }
         }
     }
 
+    private fun isShouldBeFixed(
+        condition: ASTNode,
+        elseCodeLines: List<String>?,
+        numberOfStatementsInElseBlock: Int,
+        isAssignment: Boolean,
+    ): Boolean = when {
+        // else { "null"/empty } -> ""
+        isNullOrEmptyElseBlock(elseCodeLines) -> true
+        // else { bar() } -> ?: bar()
+        isOnlyOneNonAssignmentStatementInElseBlock(numberOfStatementsInElseBlock, isAssignment) -> true
+        // else { ... } -> ?: run { ... }
+        else -> isNotInsideWrongBlock(condition)
+    }
+
+    private fun isOnlyOneNonAssignmentStatementInElseBlock(numberOfStatementsInElseBlock: Int, isAssignment: Boolean) = numberOfStatementsInElseBlock == 1 && !isAssignment
+
+    private fun isNullOrEmptyElseBlock(elseCodeLines: List<String>?) = elseCodeLines == null || elseCodeLines.singleOrNull() == "null"
+
+    private fun isNotInsideWrongBlock(condition: ASTNode): Boolean = condition.parents().none {
+        it.elementType == CLASS_INITIALIZER ||
+                (it.elementType == CALL_EXPRESSION &&
+                        it.findChildByType(REFERENCE_EXPRESSION)?.text in listOf("run", "with", "apply"))
+    }
+
     @Suppress("UnsafeCallOnNullableType")
     private fun isFixable(condition: ASTNode,
```

Comment:
Done. We need 2 functions, because one checks whether it is necessary to warn about null-check, and the second checks whether null-check can be auto fixed.